cli/vscodeipc/vscodeipc.go

```diff
@@ -0,0 +1,298 @@
+	}
+	// We need this to track upload and download!
+	options.EnableTrafficStats = true
+
+	agentConn, err := client.DialWorkspaceAgent(ctx, agentID, options)
+	if err != nil {
+		return nil, nil, err
+	}
+	api := &api{
+		agentConn: agentConn,
+	}
+	r := chi.NewRouter()
+	// This is to prevent unauthorized clients on the same machine from executing
+	// requests on behalf of the workspace.
+	r.Use(func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			token := r.Header.Get("Coder-Session-Token")
+			if token == "" {
+				httpapi.Write(r.Context(), w, http.StatusUnauthorized, codersdk.Response{
+					Message: "A session token must be provided in the `Coder-Session-Token` header.",
+				})
+				return
+			}
+			if token != client.SessionToken() {
+				httpapi.Write(r.Context(), w, http.StatusUnauthorized, codersdk.Response{
+					Message: "The session token provided doesn't match the one used to create the client.",
+				})
+				return
+			}
+			w.Header().Set("Access-Control-Allow-Origin", "*")
+			h.ServeHTTP(w, r)
+		})
+	})
+	r.Get("/port/{port}", api.port)
+	r.Get("/network", api.network)
+	r.Post("/execute", api.execute)
+	return r, api, nil
+}
+
+type api struct {
+	agentConn     *codersdk.AgentConn
+	sshClient     *ssh.Client
+	sshClientErr  error
+	sshClientOnce sync.Once
+
+	lastNetwork time.Time
+}
+
+func (api *api) Close() error {
+	if api.sshClient != nil {
+		api.sshClient.Close()
+	}
+	return api.agentConn.Close()
+}
+
+type NetworkResponse struct {
+	P2P              bool               `json:"p2p"`
+	Latency          float64            `json:"latency"`
+	PreferredDERP    string             `json:"preferred_derp"`
+	DERPLatency      map[string]float64 `json:"derp_latency"`
+	UploadBytesSec   int64              `json:"upload_bytes_sec"`
+	DownloadBytesSec int64              `json:"download_bytes_sec"`
+}
+
+// port accepts an HTTP request to dial a port on the workspace agent.
+// It uses an HTTP connection upgrade to transfer the connection to TCP.
+func (api *api) port(w http.ResponseWriter, r *http.Request) {
+	port, err := strconv.Atoi(chi.URLParam(r, "port"))
+	if err != nil {
+		httpapi.Write(r.Context(), w, http.StatusBadRequest, codersdk.Response{
+			Message: "Port must be an integer!",
+		})
+		return
+	}
+	remoteConn, err := api.agentConn.DialContext(r.Context(), "tcp", fmt.Sprintf("127.0.0.1:%d", port))
+	if err != nil {
+		httpapi.InternalServerError(w, err)
+		return
+	}
+	defer remoteConn.Close()
+
+	// Upgrade an switch to TCP!
+	w.Header().Set("Connection", "Upgrade")
+	w.Header().Set("Upgrade", "tcp")
+	w.WriteHeader(http.StatusSwitchingProtocols)
+
+	hijacker, ok := w.(http.Hijacker)
+	if !ok {
+		httpapi.InternalServerError(w, xerrors.Errorf("unable to hijack connection: %T", w))
+		return
+	}
+
+	localConn, brw, err := hijacker.Hijack()
+	if err != nil {
+		httpapi.InternalServerError(w, err)
+		return
+	}
+
+	_ = brw.Flush()
+	defer localConn.Close()
```

Comment:
Nit: Creation and closure handling as one block/unit (to avoid logic in between).

```suggestion
	defer localConn.Close()
	
	_ = brw.Flush()
```